[Driver] [MFD] [GPIO] Add ADP5585 MFD Driver

CkovMk opened the pull request:
ADP5585 is an I2C keypad scanner / GPIO expander / PWM generator MFD device from Analog Devices.

NXP i.MX 93 EVK uses it as GPIO expander. This driver enables GPIO configure, read/write and interrupt for the GPIO expander, and preserve the ablilty to support full feature of the chip.

CkovMk commented:
@zephyrbot should label "platform:NXP"

aasinclair reviewed (changes requested):
There are lots of compliance errors in this PR.
I suggest using clang-format to fix these.

aasinclair commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,489 @@
+		if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			data->output &= ~BIT(pin);
+			if (ret == 0) {
+				ret = i2c_burst_write_dt(&parent_cfg->i2c_bus,
+							 ADP5585_GPO_DATA_OUT_A,
+							 (uint8_t *)&data->output, 2U);
```
Burst write is unreliable, so please use i2c_write instead (See #20154).
You are also assuming endianness here when casting data->output (also applies to a few other places).

CkovMk replied:
I've avoided using i2c_burst_xxx apis. endianness is either: converted manually,  or write only the changed byte.

aasinclair commented on drivers/mfd/mfd_adp5585.c:
```diff
@@ -0,0 +1,143 @@
+	const struct mfd_adp5585_config *config = dev->config;
+	int ret = 0;
+
+	/** Set CONFIG to gpio by default */
```
You could write both of these registers in a single transaction.

CkovMk replied:
but "i2c_burst_xx" apis are unstable, as you stated.

aasinclair replied:
You can use i2c_write like this:

```suggestion
	/** Set CONFIG to gpio by default */
	uint8_t buf[] = { ADP5585_PIN_CONFIG_A, 0, 0 };
	i2c_write_dt(&config->i2c_bus, buf, sizeof(buf));
```

aasinclair commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,489 @@
+
+	k_sem_take(&parent_data->lock, K_FOREVER);
+
+	/** Set RPULL to high-z by default */
```
These could be a single write

CkovMk commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,489 @@
+		reg_value = adp5585_pin_drive_pp << bank_pin;
+	}
+	ret = i2c_reg_update_byte_dt(&parent_cfg->i2c_bus, ADP5585_GPO_OUT_MODE_A + bank,
+				     (0b1U << bank_pin), reg_value);
```
resolved.

CkovMk commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,489 @@
+	uint8_t reg_value;
+
+	/* ADP5585 has non-contiguous gpio pin layouts, account for this */
+	for (int i = 0; i < cfg->gap_count; i++) {
+		if ((pin >= cfg->pin_gaps[i].start) &&
+		    (pin < (cfg->pin_gaps[i].start + cfg->pin_gaps[i].len))) {
+			/* Pin is reserved unavailable */
+			LOG_ERR("pin %d is invalid for this device", pin);
+			return -ENOTSUP;
+		}
+	}
```
That's ture if pin_gaps remains constant, which is the current case. But as an MFD, adp5585 is expected to assign part of its pins to keyboard_scanner or pwm generator. I intended to reserve this for future use.
what's your opinion?

aasinclair replied:
Nice idea.
Looking around the `GPIO_DT_RESERVED_RANGES_NGPIOS` macro may do what you want:
https://github.com/zephyrproject-rtos/zephyr/blob/93af8754ecc0d0b19b55b3b2f7cb0f5e61e57284/include/zephyr/drivers/gpio.h#L566

CkovMk replied:
fixed. thanks

JiafeiPan commented on dts/bindings/gpio/adi,adp5585-gpio.yaml:
```diff
@@ -0,0 +1,30 @@
+    const: 2
+
+  ngpios:
+    const: 10
```
ngpios should be 13.

JiafeiPan commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -82,18 +82,40 @@
+			compatible = "adi,adp5585-gpio";
+			gpio-controller;
+			#gpio-cells = <2>;
+			ngpios = <10>;
```
As there are three reserved gpios, so ngpios should be 13.
Add annotation here like:  Onchip GPIO R0~R4 are gpio0~gipo4 and onchip GPIO C0~C4 are gipo8~gpio12

JiafeiPan commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,478 @@
+			/* reg_value for ADP5585_GPO_OUT_MODE */
+			reg_value = (uint8_t)(data->output >> 8);
+		}
+		ret = i2c_reg_write_byte_dt(&parent_cfg->i2c_bus,
```
In fact we can define some hardware access APIs in MFD drivers, so in this driver call MFD APIs but not access i2c directly.

JiafeiPan replied:
I mean we can define some common API in mfd driver so that any function driver of this mfd device can reuse them, maybe you can refer to driver gpio/gpio_axp192.c and mfd/mfd_axp192.c. anyway, this is related to refine the driver, current implement is also okay for me. but at least you need to fix code compatible checking issues.

aasinclair replied:
As you are using standard i2c functions directly, this is fine.
You only need to use mfd wrappers if it uses an unusual I2C protocol (e.g. 16 bit addressing), or supports multiple transports.

CkovMk replied:
then I'll leave this one as-is.

CkovMk commented on drivers/gpio/Kconfig.adp5585_gpio:
```diff
@@ -0,0 +1,20 @@
+	default y
+	depends on DT_HAS_ADI_ADP5585_GPIO_ENABLED
+	select MFD
+	select I2C
```
yes, this is an i2c device

aasinclair commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,478 @@
+LOG_MODULE_REGISTER(adp5585_gpio, CONFIG_GPIO_LOG_LEVEL);
+
+#define ADP5585_BANK(offs) (offs >> 3)
+#define ADP5585_BIT(offs)  (offs & 0b0111)
```
Please avoid binary literals

aasinclair commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,478 @@
+
+	k_sem_take(&parent_data->lock, K_FOREVER);
+
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
```
This condition already causes function exit at line 93

CkovMk replied:
Oh, line 93 was a mistate... thanks for pointing out.

aasinclair commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,478 @@
+	} else if ((flags & GPIO_INPUT) != 0) {
+		/* reg_value for ADP5585_GPIO_DIRECTION */
+		reg_value = adp5585_pin_output << bank_pin;
+	} else {
```
This condition has an explicit check earlier

CkovMk replied:
sure, remove it here.

aasinclair commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,478 @@
+	if (ret) {
+		goto out;
+	}
+	input_data = sys_le16_to_cpu(*((uint16_t*)(gpi_status_buf + 1)));
```
I think it would be clearer to have separate buffers for input and output.

CkovMk commented on drivers/gpio/gpio_adp5585.c:
```diff
@@ -0,0 +1,478 @@
+		goto out;
+	}
+
+	{
```
no, they are just for indentation. will remove them.

CkovMk commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -83,18 +83,44 @@
-&lpi2c2{
-	status = "disabled";
+&lpi2c2 {
+	status = "okay";
```
that would lead to ci build failure...

JiafeiPan replied:
okay, but is there any test case for adp5585 gpio?

CkovMk replied:
I can see gpio tests are built successfully during ci, they should have compiled the adp5585 driver without issue. But because there's no test gpio configured in the dt overlay, there's no way we could run the test against real hardware. All the test does is to test the built-in rgpio.

So the question is, what is the typical approach to add test to these gpio expanders?

CkovMk committed:
board: mimx93_evk_a55: enable mfd_adp5585 and adp5585_gpio driver.

Enable ADP5585 as GPIO expander on i.MX93 EVK.
- Add dts node for ADP5585 under lpi2c2.
- Add gpio-hog for EXP_SEL signal.
- Disable lpi2c2 and adp5585 node by default

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

dleach02 commented:
@CkovMk please resolve the merge conflicts.

@aasinclair can you return to the PR to determine if your requests for change have been addressed

kartben commented on drivers/mfd/Kconfig.adp5585:
```diff
@@ -0,0 +1,17 @@
+# Copyright 2024 NXP
+# SPDX -License-Identifier: Apache-2.0
```
```suggestion
# SPDX-License-Identifier: Apache-2.0
```

CkovMk replied:
Resolved. Several other files have this same issue. You may open an issue to track it.

kartben replied:
thanks for the quick turnaround! and done #72454 :)

CkovMk committed:
drivers: mfd: add new driver "mfd_adp5585"

Add mfd_adp5585 and gpio_adp5585 driver. This driver enables ADP5585
as an GPIO expander.
This chip is used as an GPIO expander on i.MX93 EVK. GPIO pinctrl,
read/write and interrupt is supported.
Note that ADP5585 has 2 GPIO banks with 5 pins each. The driver combines
two group into a 16-bit port. Index 0~4 correspond to R0~R4 lines, index
8~12 correspond to C0~C4 lines. Index 5~7 is reserved unavailable.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

dleach02 commented:
@carlocaione can you provide your review so we can merge this PR

dleach02 commented:
I'm not sure who to assign this to. @aasinclair is a maintainer of mfd and has previously approved this. I'm going to move this to me but if @aasinclair can give another pass since his approval was dropped after addressing a comment.
